aten/src/ATen/native/cuda/DepthwiseConv2d.cu

```diff
@@ -1,29 +1,596 @@
+    } else {
+      conv_depthwise2d_backward_kernel<0, 0><<<grid, block, 0, stream>>>(
+          grad_output_a, grad_input_a, weight_a, n, inputChannels, depthwiseMultiplier, outputChannels, width,
+          height, outputWidth, outputHeight, kW, kH, dW, dH, padW, padH, dilationW, dilationH);
+      C10_CUDA_KERNEL_LAUNCH_CHECK();
+    }
+  });
+}
+
+// Crude benchmarks suggest 256 is better than 512 and 1024
+// TODO: Autotune/use better heuristics, improve speed more.
+int getGradParamsNumThreads(int batchSize) {
+  //warp per item in a batch, up to a maximum
+  constexpr int MAX_BLOCK_SIZE = 256;
+  return std::min(batchSize * C10_WARP_SIZE, MAX_BLOCK_SIZE);
+}
+
+void conv_depthwise2d_grad_weight_out(
+                  const Tensor &input,
+                  const Tensor &grad_output,
+                  const Tensor &grad_weight,
+                  const int kW, const int kH,
+                  const int dW, const int dH,
+                  const int padW, const int padH,
+                  const int dilationW, const int dilationH) {
+  // Only handle 4D Input Tensors for now
+  TORCH_CHECK(input.numel() > 0 && input.dim() == 4);
+  TORCH_CHECK(grad_output.numel() > 0 && grad_output.dim() == 4);
+
+  // Minimal shape checking as above
+  // Same # of elements in batch
+  TORCH_CHECK(input.sizes()[0] == grad_output.sizes()[0]);
+
+  auto in_sizes = input.sizes();
+  int batchSize = in_sizes[0];
+  int inputChannels = in_sizes[1];
+  int height = in_sizes[2];
+  int width = in_sizes[3];
+
+  auto gO_sizes = grad_output.sizes();
+  int outputChannels = gO_sizes[1];
+  int outputHeight = gO_sizes[2];
+  int outputWidth = gO_sizes[3];
+
+  int depthwiseMultiplier = outputChannels / inputChannels;
+
+  resize_output(grad_weight, {outputChannels, 1, kH, kW});
+
+  // Kernel currently relies upon all the Tensors to be contiguous
+  TORCH_CHECK(grad_output.is_contiguous());
+  TORCH_CHECK(input.is_contiguous());
+  TORCH_CHECK(grad_weight.is_contiguous());
+
+  // We parallelize so that each block computes a single value in grad_weight
+  TORCH_CHECK(canUse32BitIndexMath(input) &&
+              canUse32BitIndexMath(grad_output));
+  int blocks = outputChannels * kH * kW;
+
+  // Make sure we have enough threads to perform the reduction, and use this number
+  // to create the shared memory size for the reduction
+  dim3 grid(blocks);
+  dim3 block(getGradParamsNumThreads(batchSize));
+  const auto stream = c10::cuda::getCurrentCUDAStream();
+
+  AT_DISPATCH_FLOATING_TYPES_AND2(kHalf, kBFloat16, grad_output.scalar_type(),
+                                  "conv_depthwise2d_grad_weight_cuda", [&] {
+    const auto grad_output_a = grad_output.packed_accessor32<scalar_t, 4>();
+    const auto input_a = input.packed_accessor32<scalar_t, 4>();
+    const auto grad_weight_a = grad_weight.packed_accessor32<scalar_t, 4>();
+    using acc_t = at::acc_type<scalar_t, true>;
+    int smem = ((block.x + C10_WARP_SIZE - 1) / C10_WARP_SIZE) * sizeof(acc_t);
```

Comment:
I think blockReduceSum cannot handle non-multiple-of-warp number of threads, so this can be changed back to `int smem = block.x * sizeof(acc_t)`, and we can add TORCH_INTERNAL_ASSERT that (block.x % C10_WARP_SIZE ==0)